src/command/supergraph/compose.rs

```diff
@@ -43,3 +61,197 @@ impl Compose {
+                        schema_path
+                    }
+                    None => file.clone(),
+                };
+
+                let schema = fs::read_to_string(&relative_schema_path).map_err(|e| {
+                    let err = anyhow!("Could not read \"{}\": {}", &relative_schema_path, e);
+                    let mut err = RoverError::new(err);
+                    err.set_suggestion(Suggestion::ValidComposeFile);
+                    err
+                })?;
+
+                let url = &subgraph_data.routing_url.clone().ok_or_else(|| {
+                    let err = anyhow!("No routing_url found for schema file.");
+                    let mut err = RoverError::new(err);
+                    err.set_suggestion(Suggestion::ValidComposeRoutingUrl);
+                    err
+                })?;
+
+                let subgraph_definition = SubgraphDefinition::new(subgraph_name, url, &schema);
+                subgraphs.push(subgraph_definition);
+            }
+            SchemaSource::SubgraphIntrospection { subgraph_url } => {
+                // given a federated introspection URL, use subgraph introspect to
+                // obtain SDL and add it to subgraph_definition.
+                let client = Client::new(&subgraph_url.to_string());
+
+                let introspection_response = introspect::run(&client, &HashMap::new())?;
+                let schema = introspection_response.result;
+
+                let subgraph_definition = SubgraphDefinition::new(subgraph_name, "", &schema);
+                subgraphs.push(subgraph_definition);
+            }
+            SchemaSource::Introspection { url } => {
+                // given an introspection URL, use graph introspect to
+                // obtain SDL and add it to subgraph_definition.
+                let client = Client::new(&url.to_string());
+
+                let introspection_response = graph::introspect::run(&client, &HashMap::new())?;
+                let schema = introspection_response.result;
+
+                let subgraph_definition = SubgraphDefinition::new(subgraph_name, "", &schema);
+                subgraphs.push(subgraph_definition);
+            }
+            SchemaSource::Subgraph { graphref, subgraph } => {
+                // given a graphref and subgraph, run subgraph fetch to
+                // obtain SDL and add it to subgraph_definition.
+                let client = client_config.get_client(&profile_name)?;
+                let graphref = parse_graph_ref(graphref)?;
+                let schema = fetch::run(
+                    fetch::fetch_subgraph_query::Variables {
+                        graph_id: graphref.name.clone(),
+                        variant: graphref.variant.clone(),
+                    },
+                    &client,
+                    subgraph,
+                )?;
+
+                let subgraph_definition = SubgraphDefinition::new(subgraph_name, "", &schema);
+                subgraphs.push(subgraph_definition);
```

Comment:
I think we're only currently getting the `name` and the `sdl` from Studio, and not also getting the `url` (which is the `routing_url`).  And I _believe_ this is happening via `fetch_subgraph_query` and the omission of the `url` is in this `fetch.graphql` file.  (Trust nothing that I'm saying though, these are only beliefs and I'm not all tote familiar with how the GraphQL client works here.)

https://github.com/apollographql/rover/blob/9ad331a9e462bd40ef3daa6b26952d2f8c33a842/crates/rover-client/src/query/subgraph/fetch.graphql#L6-L11

Should we also get that or is the intention that we require users to always specify `routing_url` locally?

I do see that this suggestion would require broader changes to the `query::subgraph::fetch::run` method to allow it to return something more complex than merely the `String` of SDL though.  I would claim that we can probably punt on this, though we may want to just _require_ the presence of `routing_url` in the YAML when using a `graphref` fetch?

Reply:
As discussed in slack, let's always use the routing_url if provided, and otherwise supply an empty string to harmonizer to run composition. In the future, we can also tweak the request sent to studio and use the routing_url being used in studio.